### DISCUSSION TITLE
Problem encountered with HTML

### DISCUSSION THREAD

**Discussion by ybenarbiaBBI** [clientIO/joint#1776] [Q&A] [answered]
Hello everyone,

I have a problem with the displaying of my Hello world components with the free version of JointJS in my angular component HTML page.

as you can see, the code below isn't working :

![pb1](https://user-images.githubusercontent.com/99811254/183875080-da355cd1-2063-4465-ac20-f5a79b1adbe4.PNG)
![pb2](https://user-images.githubusercontent.com/99811254/183875268-eead8e53-feff-4c15-849a-64cf3c02424e.PNG)

So please if you've an idea in how to fix that, tell me.

Thank for your response guys !

**Comment by jamesgeorgewilliams** [comment, marked as answer]
Just from looking at `script` tags in your picture, I can see you are not using the latest version of JointJS via CDN.

The latest CDN versions can be seen in this tutorial:
https://resources.jointjs.com/tutorial/hello-world

If you have some errors in the console, it might help you debug the problem.

The following tutorial/repository shows you how to use JointJS+ with Angular, but you could create similar code using JointJS.
https://resources.jointjs.com/tutorial/angular
https://github.com/clientIO/rappid-tutorial-angular12

**Reply by ybenarbiaBBI** [reply]
Hi, thank for your answer,

I've put instead the last versions of CDN but there is no display yet.
Unfortunately, i have none errors in the console that might help me.

For jointJS+ with angular i can't work with because i can't get the license.

**Reply by jamesgeorgewilliams** [reply]
But you can install `jointjs` via npm, and use it in the same manner as the tutorial above:
https://www.npmjs.com/package/jointjs

I don't think your project structure is correct, and installing via npm is probably the way to go if creating components like you are doing in the pictures above. 

The code above is a demonstration that should be placed in an `index.html` file.

**Reply by ybenarbiaBBI** [reply]
thank it's working 

but is there a method to display html content directly on an app.component.html file ?

i've tested it but that doesn't works

**Reply by jamesgeorgewilliams** [reply]
JointJS is a JavaScript library. If you want to create JointJS elements, you have to do so programmatically in JavaScript.

It is not possible to create JointJS elements via html.

If you want to create JointJS elements in the `app.component.html` or `AppComponent`, you have to create them in the respective TypeScript file. In this case it would be `app.component.ts`.

The following is an example for Angular.
https://github.com/clientIO/rappid-tutorial-angular12/blob/main/src/app/app.component.ts

**Reply by ybenarbiaBBI** [reply]
I see, thank you for the clearification once more !

but now i have an other error who's appearing with this message :

Error: src/app/app.component.ts:2:33 - error TS2307: Cannot find module '@clientio/rappid' or its corresponding type declarations.

2 import { dia, ui, shapes } from '@clientio/rappid';
                                  ~~~~~~~~~~~~~~~~~~


Error: src/app/app.component.ts:10:24 - error TS2564: Property 'canvas' has no initializer and is not definitely assigned in the constructor.

10   @ViewChild('canvas') canvas: ElementRef;



i've already tried to change the imports by searching but i can't see from where's the problem..

**Reply by jamesgeorgewilliams** [reply]
Yes, as stated above the linked repository uses JointJS+, so you need to adapt the code for JointJS.

The linked repository contains code for the `PaperScroller` plugin which is a JointJS+ plugin, so that is one thing you need to remove for example.


When using JointJS and Angular to follow the `Hello World!` tutorial, the files could look like the following:

app.component.html:
```html
<div id="myholder"></div>
```

app.component.ts:
```js
import { OnInit, Component } from '@angular/core';
import { dia, shapes } from 'jointjs';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit {

  private graph: dia.Graph;
  private paper: dia.Paper;

  public ngOnInit(): void {

    const namespace = shapes;

    const graph = new dia.Graph({}, { cellNamespace: namespace });;

    const paper = new dia.Paper({
      el: document.getElementById('myholder')!,
      model: graph,
      width: 600,
      height: 100,
      gridSize: 1,
      sorting: dia.Paper.sorting.APPROX,
      cellViewNamespace: namespace
    });

    const rect = new shapes.standard.Rectangle();
    rect.position(100, 30);
    rect.resize(100, 40);
    rect.attr({
        body: {
            fill: 'blue'
        },
        label: {
            text: 'Hello',
            fill: 'white'
        }
    });
    rect.addTo(graph);

    const rect2 = rect.clone();
    rect2.translate(300, 0);
    rect2.attr('label/text', 'World!');
    rect2.addTo(graph);

    const link = new shapes.standard.Link();
    link.source(rect);
    link.target(rect2);
    link.addTo(graph);
  }
}

```

You still need to follow various steps in your Angular project as outlined in the tutorial.

To setup styles, add the following to `angular.json`:
```js
"styles": [
     "src/styles.css",
     "node_modules/jointjs/dist/joint.css"
 ],
```

You will need to add some `@types` to the `package.json` as mentioned in the tutorial, and install them via `npm`.
```js
"@types/backbone": "latest",
"@types/jquery": "latest",
```
 You should also update the `tsconfig.json` with the following rules:
 ```js
"strictPropertyInitialization": false,
"skipLibCheck": true,
 ```

I suggest studying some of our tutorials to understand everything.

It is important to understand what features are part of JointJS and JointJS+, so you can adapt the code accordingly.

**Reply by ybenarbiaBBI** [reply]
thank you, i see

i'll try to do that

so about PaperScroller  plug-in did you know how to remove this one from  the joint repository ?

**Reply by jamesgeorgewilliams** [reply]
Just copy & paste the above code to your Angular project in the correct files, and it should work.

The code above doesn't contain the `PaperScroller` plugin, because it is using the JointJS library only.

**Reply by ybenarbiaBBI** [reply]
Ok i see,  thank you !

Yes I've done that but when i run the project, the terminal give me that error now :


Build at: 2022-08-11T10:04:55.483Z - Hash: 130d2eba3fb4eadc - Time: 14472ms

Warning: C:\Users\y.benarbia\testultime\node_modules\jointjs\src\mvc\index.mjs depends on 'backbone'. CommonJS or AMD dependencies can cause optimization bailouts.
For more info see: https://angular.io/guide/build#configuring-commonjs-dependencies

Warning: C:\Users\y.benarbia\testultime\node_modules\jointjs\src\mvc\index.mjs depends on 'jquery'. CommonJS or AMD dependencies can cause optimization bailouts.
For more info see: https://angular.io/guide/build#configuring-commonjs-dependencies

Warning: C:\Users\y.benarbia\testultime\node_modules\jointjs\src\util\util.mjs depends on 'lodash'. CommonJS or AMD dependencies can cause optimization bailouts.
For more info see: https://angular.io/guide/build#configuring-commonjs-dependencies



Error: Module not found: Error: Can't resolve 'C:/Users/y.benarbia/testultime/src/app/app.component.css?ngResource' in 'C:\Users\y.benarbia\testultime'

Error: The loader "C:/Users/y.benarbia/testultime/src/app/app.component.css" didn't return a string.



** Angular Live Development Server is listening on localhost:51938, open your browser on http://localhost:51938/ **


× Failed to compile.

**Reply by ybenarbiaBBI** [reply]
sorryy, i say nothing haha it's working, thank you so much for your help !!!!! i have just to replace 
styleUrls: ['./app.component.css'] 

by :

styleUrls: ['./app.component.scss']

in my app.component.ts
